google-beta/resource_pubsub_subscription.go

```diff
@@ -63,166 +77,375 @@ func resourcePubsubSubscription() *schema.Resource {
+		d.Set("topic", fullTopic)
+		return fullTopic, nil
+	}
+}
+
+func expandPubsubSubscriptionLabels(v interface{}, d *schema.ResourceData, config *Config) (map[string]string, error) {
+	if v == nil {
+		return map[string]string{}, nil
+	}
+	m := make(map[string]string)
+	for k, val := range v.(map[string]interface{}) {
+		m[k] = val.(string)
+	}
+	return m, nil
 }
 
-func flattenPubsubSubscriptionPushConfig(pushConfig *pubsub.PushConfig) []map[string]interface{} {
-	configs := make([]map[string]interface{}, 0, 1)
+func expandPubsubSubscriptionPushConfig(v interface{}, d *schema.ResourceData, config *Config) (interface{}, error) {
+	l := v.([]interface{})
+	if len(l) == 0 || l[0] == nil {
+		return nil, nil
+	}
+	raw := l[0]
+	original := raw.(map[string]interface{})
+	transformed := make(map[string]interface{})
+
+	transformedPushEndpoint, err := expandPubsubSubscriptionPushConfigPushEndpoint(original["push_endpoint"], d, config)
+	if err != nil {
+		return nil, err
+	} else if val := reflect.ValueOf(transformedPushEndpoint); val.IsValid() && !isEmptyValue(val) {
+		transformed["pushEndpoint"] = transformedPushEndpoint
+	}
 
-	if pushConfig == nil || len(pushConfig.PushEndpoint) == 0 {
-		return configs
+	transformedAttributes, err := expandPubsubSubscriptionPushConfigAttributes(original["attributes"], d, config)
+	if err != nil {
+		return nil, err
+	} else if val := reflect.ValueOf(transformedAttributes); val.IsValid() && !isEmptyValue(val) {
+		transformed["attributes"] = transformedAttributes
 	}
 
-	configs = append(configs, map[string]interface{}{
-		"push_endpoint": pushConfig.PushEndpoint,
-		"attributes":    pushConfig.Attributes,
-	})
+	return transformed, nil
+}
+
+func expandPubsubSubscriptionPushConfigPushEndpoint(v interface{}, d *schema.ResourceData, config *Config) (interface{}, error) {
+	return v, nil
+}
+
+func expandPubsubSubscriptionPushConfigAttributes(v interface{}, d *schema.ResourceData, config *Config) (map[string]string, error) {
+	if v == nil {
+		return map[string]string{}, nil
+	}
+	m := make(map[string]string)
+	for k, val := range v.(map[string]interface{}) {
+		m[k] = val.(string)
+	}
+	return m, nil
+}
 
-	return configs
+func expandPubsubSubscriptionAckDeadlineSeconds(v interface{}, d *schema.ResourceData, config *Config) (interface{}, error) {
+	return v, nil
 }
 
-func expandPubsubSubscriptionPushConfig(configured []interface{}) *pubsub.PushConfig {
-	if len(configured) == 0 || configured[0] == nil {
-		// An empty `pushConfig` indicates that the Pub/Sub system should stop pushing messages
-		// from the given subscription and allow messages to be pulled and acknowledged.
-		return &pubsub.PushConfig{}
+func resourcePubsubSubscriptionUpdateEncoder(d *schema.ResourceData, meta interface{}, obj map[string]interface{}) (map[string]interface{}, error) {
+	newObj := make(map[string]interface{})
+	newObj["subscription"] = obj
+	updateMask := make([]string, 0)
```

Comment:
Are we able to mark the resource with https://github.com/GoogleCloudPlatform/magic-modules/blob/fb5e56f056cdfe608196b00d36bc0cdf434138d1/api/resource.rb#L182 in `api.yaml` and reduce this encoder to exclusively put the object under a `subscription` JSON field?